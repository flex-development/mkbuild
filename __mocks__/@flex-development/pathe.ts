/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
