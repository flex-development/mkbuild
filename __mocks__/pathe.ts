/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
