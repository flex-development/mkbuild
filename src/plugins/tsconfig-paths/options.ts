/**
 * @file Plugin Options - tsconfig-paths
 * @module mkbuild/plugins/tsconfig-paths/options
 */

import type { ReadJsonSync } from 'tsconfig-paths/lib/filesystem'

/**
 * [`tsconfig-paths`][1] plugin options.
 *
 * [1]: https://github.com/dividab/tsconfig-paths
 */
interface TsconfigPathsPluginOptions {
  /**
   * Module extensions to probe for.
   *
   * @default MODULE_EXTENSIONS
   */
  extensions?: string[]

  /**
   * Checks for the existence of a file at `path`.
   *
   * @param {string} path - Path to check
   * @return {boolean} `true` if file exists, `false` otherwise
   */
  fileExists?(path: string): boolean

  /**
   * `package.json` fields to check when resolving modules.
   *
   * A nested field can be selected by passing an array of field names.
   *
   * @default ['main', 'module']
   */
  mainFields?: (string[] | string)[]

  /**
   * Reads `JSON` data from a file.
   */
  readJson?: ReadJsonSync
}

export type { TsconfigPathsPluginOptions as default }
