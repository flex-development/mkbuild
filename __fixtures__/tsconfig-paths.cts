/**
 * @file Fixtures - tsconfig-paths
 * @module fixtures/tsconfig-paths.cts
 */

require('#src/config/constants')
require('#src/interfaces')
require('emoji-regex')
require('esbuild')
require('pathe')
require('tsconfig-paths')
require('tsconfig-paths/lib/tsconfig-loader')
