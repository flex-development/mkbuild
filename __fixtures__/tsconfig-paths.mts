/**
 * @file Fixtures - tsconfig-paths
 * @module fixtures/tsconfig-paths.mts
 */

import '#src/config/constants'
import '#src/interfaces'
import 'emoji-regex'
import 'esbuild'
import 'pathe'
import 'tsconfig-paths'
import 'tsconfig-paths/lib/tsconfig-loader'
