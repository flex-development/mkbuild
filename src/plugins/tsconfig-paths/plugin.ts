/**
 * @file Plugins - tsconfig-paths
 * @module mkbuild/plugins/tsconfig-paths/plugin
 */

import {
  EXT_JS_REGEX,
  EXT_TS_REGEX,
  RESOLVE_EXTENSIONS
} from '#src/config/constants'
import type { OutputMetadata } from '#src/types'
import extractStatements from '#src/utils/extract-statements'
import type {
  BuildOptions,
  BuildResult,
  OutputFile,
  Plugin,
  PluginBuild
} from 'esbuild'
import * as pathe from 'pathe'
import { createMatchPath, type MatchPath } from 'tsconfig-paths'
import {
  tsConfigLoader,
  type TsConfigLoaderResult
} from 'tsconfig-paths/lib/tsconfig-loader'
import type Options from './options'

/**
 * Plugin name.
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'tsconfig-paths'

/**
 * Returns a [`tsconfig-paths`][1] plugin.
 *
 * [1]: https://github.com/dividab/tsconfig-paths
 *
 * @param {Options} [options={}] - Plugin options
 * @return {Plugin} `tsconfig-paths` plugin
 */
const plugin = ({
  extensions = RESOLVE_EXTENSIONS,
  fileExists,
  mainFields = ['main', 'module'],
  readJson
}: Options = {}): Plugin => {
  /**
   * Resolves path aliases in output file content.
   *
   * [1]: https://github.com/dividab/tsconfig-paths
   * [2]: https://esbuild.github.io/plugins
   * [3]: https://esbuild.github.io/api/#build-api
   *
   * @param {PluginBuild} build - [esbuild plugin api][2]
   * @param {BuildOptions} build.initialOptions - [esbuild build api][3] options
   * @param {PluginBuild['onEnd']} build.onEnd - Build end callback
   * @return {void} Nothing when complete
   * @throws {Error}
   */
  const setup = ({ initialOptions, onEnd }: PluginBuild): void => {
    const {
      absWorkingDir = process.cwd(),
      bundle,
      metafile,
      tsconfig
    } = initialOptions

    // esbuild handles path aliases when bundling
    if (bundle) return

    // metafile required to get output metadata
    if (!metafile) throw new Error('metafile required')

    /**
     * Tsconfig loader result.
     *
     * @const {TsConfigLoaderResult} result
     */
    const tsconfig_result: TsConfigLoaderResult = tsConfigLoader({
      cwd: absWorkingDir,
      /**
       * Returns the value of `key`.
       *
       * @param {string} key - Environment variable key
       * @return {string | undefined} Value of `key`
       */
      getEnv(key: string): string | undefined {
        return key === 'TS_NODE_PROJECT'
          ? tsconfig ?? process.env[key]
          : process.env[key]
      }
    })

    // tsconfig couldn't be found
    if (!tsconfig_result.tsConfigPath) throw new Error('tsconfig not found')

    // tsconfig data
    const { baseUrl = '', paths = {}, tsConfigPath } = tsconfig_result

    /**
     * Path matching function.
     *
     * @const {MatchPath} matcher
     */
    const matcher: MatchPath = createMatchPath(
      pathe.resolve(pathe.dirname(tsConfigPath), baseUrl),
      paths,
      mainFields,
      baseUrl.length > 0
    )

    return void onEnd((result: BuildResult): void => {
      result.outputFiles = result.outputFiles!.map((output: OutputFile) => {
        /**
         * Output file extension.
         *
         * @const {string} output_ext
         */
        const output_ext: string = pathe.extname(output.path)

        // skip output files that are not javascript or typescript
        if (!EXT_JS_REGEX.test(output_ext) && !EXT_TS_REGEX.test(output_ext)) {
          return output
        }

        /**
         * Relative path to output file.
         *
         * **Note**: Relative to {@link absWorkingDir}.
         *
         * @const {string} outfile
         */
        const outfile: string = output.path.replace(absWorkingDir + '/', '')

        /**
         * {@link output} metadata.
         *
         * @const {OutputMetadata} metadata
         */
        const metadata: OutputMetadata = result.metafile!.outputs[outfile]!

        // because this plugin doesn't handle bundles, the entry point can be
        // reset to the first (and only!) key in metadata.inputs
        if (!metadata.entryPoint) {
          metadata.entryPoint = Object.keys(metadata.inputs)[0]!
        }

        /**
         * Absolute path to source file.
         *
         * @const {string} source
         */
        const source: string = pathe.resolve(absWorkingDir, metadata.entryPoint)

        /**
         * {@link output.text} copy.
         *
         * @var {string} text
         */
        let text: string = output.text

        // replace path aliases
        for (const statement of extractStatements(output.text)) {
          // do nothing if missing module specifier
          if (!statement.specifier) continue

          /**
           * Possible path match for {@link statement.specifier}.
           *
           * @var {string | undefined} match
           */
          let match: string | undefined = matcher(
            statement.specifier,
            readJson,
            fileExists,
            extensions
          )

          // do nothing if path match was not found
          if (!match) continue

          // remove node_modules reference or set match relative to source dir
          match = /\/node_modules\//.test(match)
            ? match.replace(/.+\/node_modules\//, '')
            : pathe
                .relative(pathe.dirname(source), match)
                .replace(/^(\w)/, './$1')
                .replace(/(\/index)$/, '')

          // replace path alias
          text = text.replace(
            statement.code,
            statement.code.replace(statement.specifier, match)
          )
        }

        return { ...output, contents: new Uint8Array(Buffer.from(text)), text }
      })
    })
  }

  return { name: PLUGIN_NAME, setup }
}

export { plugin as default, type Options }
