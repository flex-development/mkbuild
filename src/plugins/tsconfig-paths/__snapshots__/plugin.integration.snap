// Vitest Snapshot v1

exports[`integration:plugins/tsconfig-paths > esbuild > should resolve path aliases esm output files 1`] = `
"import \\"../src/config/constants\\";
import \\"../src/utils/extract-statements\\";
import \\"emoji-regex/es2015\\";
import \\"esbuild\\";
import \\"pathe\\";
import \\"tsconfig-paths\\";
import \\"tsconfig-paths/lib/tsconfig-loader\\";
const foo = \\"foo\\";
export * from \\"../src/interfaces\\";
export {
  foo
};
"
`;

exports[`integration:plugins/tsconfig-paths > esbuild > should resolve path aliases in cjs output files 1`] = `
"require(\\"../src/config/constants\\");
require(\\"../src/utils/extract-statements\\");
require(\\"emoji-regex/es2015\\");
require(\\"esbuild\\");
require(\\"pathe\\");
require(\\"tsconfig-paths\\");
require(\\"tsconfig-paths/lib/tsconfig-loader\\");
"
`;

exports[`integration:plugins/tsconfig-paths > esbuild > should resolve path aliases in copied output files 1`] = `
"declare module '__fixtures__/values' {
  export { default as MAX_VALUE } from './max-value'
  export { default as MIN_VALUE } from './min-value'
}
"
`;

exports[`integration:plugins/tsconfig-paths > esbuild > should skip files that are not javascript or typescript 1`] = `
"[
  {
    \\"date\\": \\"2007-04-24T07:00:00.000Z\\",
    \\"close\\": 93.24
  },
  {
    \\"date\\": \\"2007-04-25T07:00:00.000Z\\",
    \\"close\\": 95.35
  },
  {
    \\"date\\": \\"2007-06-08T07:00:00.000Z\\",
    \\"close\\": 124.49
  },
  {
    \\"date\\": \\"2007-06-11T07:00:00.000Z\\",
    \\"close\\": 120.19
  },
  {
    \\"date\\": \\"2007-06-12T07:00:00.000Z\\",
    \\"close\\": 120.38
  }
]
"
`;
